### PR TITLE
[Improvement]downgrade resource tag when there is not queryable replica

### PR DESCRIPTION
**Pull request by wangbo**
### What problem does this PR solve?

Issue Number: close #xxx

Related PR: #xxx

Problem Summary:
Add a user property  ```allow_resource_tag_downgrade``` ,  when ```allow_resource_tag_downgrade = true```, it means when resource tag has no a queryable replica,then resource tag is disabled and try to find a queryable replica.
In some scenarios, query success rate takes precedence over resource isolation.

currently support stream load and select.

### Release note

None

### Check List (For Author)

- Test 
    - [x] Regression test
    - [ ] Unit Test
    - [ ] Manual test (add detailed scripts or steps below)
    - [ ] No need to test or manual test. Explain why:
        - [ ] This is a refactor/code format and no logic has been changed.
        - [ ] Previous test can cover this change.
        - [ ] No code files have been changed.
        - [ ] Other reason 

- Behavior changed:
    - [x] No.
    - [ ] Yes. 

- Does this need documentation?
    - [ ] No.
    - [x] Yes. 

### Check List (For Reviewer who merge this PR)

- [ ] Confirm the release note
- [ ] Confirm test cases
- [ ] Confirm document
- [ ] Add branch pick label

### CODE REVIEW

**Comment by wangbo** [comment]
run buildall

**Comment by wangbo** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/qe/SessionVariable.java [hunk 1]
```diff
@@ -1528,6 +1530,15 @@ public void setEnableLeftZigZag(boolean enableLeftZigZag) {
     @VariableMgr.VarAttr(name = SKIP_BAD_TABLET)
     public boolean skipBadTablet = false;
 
+    @VariableMgr.VarAttr(name = SKIP_RESOURCE_TAG_BAD_REPLICA, description = {
+            "当一个resource tag内的副本都不可查询的时候，默认情况下查询会失败。如果设置当前变量为true，"
+                    + "那么至少会找到一个可用的副本，保证查询的成功，这会导致查询路由到其他resource tag。",
+            "If none of the replicas within a resource tag are available for querying, the query will fail by default. "
```
Comment (yiguolei) [inline fe/fe-core/src/main/java/org/apache/doris/qe/SessionVariable.java:1536]: Do not use name   resource tag,  use resource group  in all documents and variable names and comments.

**Comment by wangbo** [comment]
run buildall

**Comment by wangbo** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/mysql/privilege/UserProperty.java [hunk 2]
```diff
@@ -358,6 +367,15 @@ public void update(List<Pair<String, String>> properties, boolean isReplay) thro
                     throw new DdlException("workload group " + value + " not exists");
                 }
                 workloadGroup = value;
+            } else if (keyArr[0].equalsIgnoreCase(PROP_ALLOW_RESOURCE_TAG_DOWNGRADE)) {
+                if (keyArr.length != 1) {
+                    throw new DdlException(PROP_ALLOW_RESOURCE_TAG_DOWNGRADE + " format error");
+                }
+                if (!"true".equals(value) && !"false".equals(value)) {
```
Comment (yiguolei) [inline fe/fe-core/src/main/java/org/apache/doris/mysql/privilege/UserProperty.java:374]: equals ignore case？

File: fe/fe-core/src/main/java/org/apache/doris/planner/OlapScanNode.java [hunk 3]
```diff
@@ -890,6 +890,16 @@ private void addScanRangeLocations(Partition partition,
...
             int replicaInTablet = 0;
             long oneReplicaBytes = 0;
+
+
+            // when resource tag has no alive replica and allowResourceTagDowngrade = true,
+            // resource tag should be disabled, we should find at least one alive replica
+            boolean shouldSkipResourceTag = false;
+            boolean isAllowRgDowngrade = context.isAllowResourceTagDowngrade();
+            if (needCheckTags && isAllowRgDowngrade && !checkTagHasAvailReplica(allowedTags, replicas)) {
+                shouldSkipResourceTag = true;
```
Comment (yiguolei) [inline fe/fe-core/src/main/java/org/apache/doris/planner/OlapScanNode.java:900]: need log here， to print skip resource tag and the query id

File: fe/fe-core/src/test/java/org/apache/doris/planner/ResourceTagQueryTest.java [hunk 4]
```diff
@@ -280,7 +280,7 @@ public void test() throws Exception {
         Assert.assertEquals(1000000, execMemLimit);
 
         List<List<String>> userProps = Env.getCurrentEnv().getAuth().getUserProperties(Auth.ROOT_USER);
-        Assert.assertEquals(13, userProps.size());
+        Assert.assertEquals(14, userProps.size());
 
         // now :
```
Comment (yiguolei) [inline fe/fe-core/src/test/java/org/apache/doris/planner/ResourceTagQueryTest.java:285]: Should add test  for degrade logic。

File: fe/fe-core/src/main/java/org/apache/doris/system/BeSelectionPolicy.java [hunk 5]
```diff
@@ -176,7 +183,16 @@ private boolean isMatch(Backend backend) {
     }
 
     public List<Backend> getCandidateBackends(Collection<Backend> backends) {
-        List<Backend> filterBackends = backends.stream().filter(this::isMatch).collect(Collectors.toList());
+        boolean needResourceTagAvail = !this.allowResourceTagDowngrade || backends.stream()
```
Comment (yiguolei) [inline fe/fe-core/src/main/java/org/apache/doris/system/BeSelectionPolicy.java:186]: need unit test for this method

**Comment by wangbo** [comment]
run buildall

**Comment by wangbo** [comment]
run buildall

**Review by yiguolei — APPROVED** [review]

**Review by xinyiZzz — APPROVED** [review]
LGTM

**Pull request merged**